File: JavaCivitas/src/civitas/JugadorEspeculador.java
package civitas;

public class JugadorEspeculador extends Jugador{

    private static int FactorEspeculador = 2;
    private int fianza;
    protected int CasasMax = 8;
    protected int CasasPorHotel = 4;
    protected int HotelesMax = 8;

    private void actualizaPropietarioPorConversion(Jugador jugador){

        for (TituloPropiedad cada : jugador.getPropiedades()){

            cada.actualiarPropietarioPorConversion(this);

        }

    }

    JugadorEspeculador(Jugador real_jugador, int fianza_a){

        super(real_jugador);
        fianza = fianza_a;
        actualizaPropietarioPorConversion(real_jugador);

    }

    @Override
    public String toString(){

        return super.toString() + "[JUGADOR ESPECULADOR]";

    }

    @Override
    boolean pagaImpuesto(float cantidad){

        if(encarcelado){

            return false;

        }else{

            return paga(cantidad/2);

        }

    }

    @Override
    protected boolean debeSerEncarcelado(){

        if (encarcelado){

            return false;

        }else{

            if(tieneSalvoconducto()){

                perderSalvoConducto();
                Diario.getInstance().ocurreEvento("El jugador "+ getNombre() +" ha usado SALVOCONDUCTO");
                return false;
            }else{

                if(saldo > fianza){

                    paga(fianza);
                    return false;

                }

            }

        }

        return true;

    }

    protected boolean puedoEdificarCasa(TituloPropiedad prop){

        return prop.getPropietario() == this && getSaldo() >= prop.getPrecioEdificar()
                && prop.getNumCasas() > CasasPorHotel && prop.getNumHoteles() < HotelesMax;

    }

}

File: RubyCivitas/lib/jugador_especulador.rb
require_relative 'jugador'

module Civitas
  class Jugador_especulador < Jugador
    
    @@casas_max = 8
    @@casas_por_hotel = 4
    @@hoteles_max = 8
    
    def initialize(fianza)
      
      @fianza = fianza
      
    end
    
    def self.copia(otro, fianza)
      
      yo=new(fianza)
      yo.copia(otro)
      
      for i in otro.propiedades
        i.actualiza_propietario_por_conversion(yo)
      end
      
      yo
      
    end
    
    def to_string
      
      return super + "[JUGADOR ESPECULADOR]"
      
    end
    
    def paga_impuesto(cantidad)
      
      if @encarcelado
        
        return false
      else
        
        paga(cantidad/2)
        
      end
    end
    
    def construir_casa(ip)
      
      resultado = false
      
      if @encarcelado
        
        return resultado
        
      end
      
      if !@encarcelado
        
        
        existe = existe_la_propiedad(ip)
        
        
        if existe
          
          propiedad = @propiedades[ip]
          puedo_edificar = puedo_edificar_casa(propiedad)
          
          if puedo_edificar
            
            resultado = propiedad.construir_casa(self)
            
          end
          
        end
        
      end
      
      return resultado
    end
    
    def construir_hotel(ip)
      
      if !@encarcelado && existe_la_propiedad(ip)

        propiedad = @propiedades[ip]
        puedo_edificar_hotel = puedo_edificar_hotel(propiedad)

        if puedo_edificar_hotel

          result = propiedad.construir_hotel(self)

          casas_por_hotel = @@casas_por_hotel
          propiedad.derruir_casas(casas_por_hotel, self)

          Diario.instance.ocurre_evento("[DIARIO] El jugador " + @nombre + " construye un hotel " + "en " +
                                     ip + ", " + propiedad.nombre)
        end
      end

      return result
      
    end
    
    def encarcelar(num_casilla_carcel)
      
      if debe_ser_encarcelado

        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha sido encarcelado")
      end

      return @encarcelado
      
    end
    
    def debe_ser_encarcelado
      
      if @encarcelado
        return false
      else
        
        if tiene_salvoconducto
          
          perder_salvoconducto
          Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha usado SALVOCONDUCTO")
          return false
        else
          
          if @saldo > @fianza
            
            self.paga(@fianza)
            return false
            
          end
          
        end
        
      end
      
      return true
    end
    
    def salir_carcel_pagando
      
      resultado = false
      
      if puede_salir_carcel_pagando && @encarcelado
        
        paga(@fianza)
        @encarcelado = false
        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha pagado, sale de la carcel")
        resultado = true
        
      end
      
      return resultado
      
    end
    
    attr_accessor :fianza
    private_class_method :new
        
  end
end
